File: backend/app/utils/mentions.py
import re

from sqlalchemy.orm import Session

from app.models import User

MENTION_PATTERN = re.compile(r"@([\w]+(?:\s[\w]+)?)")


def resolve_mentioned_user_ids(
    content: str,
    db: Session,
    explicit_ids: list[int] | None = None,
) -> set[int]:
    """Resolve @mentions in comment text to user IDs."""
    result = set(explicit_ids or [])
    users = db.query(User).filter(User.status != "inactive").all()
    if not users:
        return result

    lookup: dict[str, int] = {}
    for user in users:
        full = user.full_name.lower()
        lookup[full] = user.id
        lookup[user.first_name.lower()] = user.id
        lookup[user.email.split("@")[0].lower()] = user.id

    for match in MENTION_PATTERN.finditer(content):
        token = match.group(1).strip().lower()
        if token in lookup:
            result.add(lookup[token])
            continue
        for full_name, user_id in lookup.items():
            if " " in full_name and (full_name == token or full_name.startswith(f"{token} ")):
                result.add(user_id)
                break

    return result

File: frontend/src/lib/mentions.ts
export type MentionUser = {
  id: number;
  first_name: string;
  last_name: string;
  email?: string;
  profile_picture?: string;
};

export function mentionDisplayName(user: MentionUser): string {
  return `${user.first_name} ${user.last_name}`;
}

export function mentionToken(user: MentionUser): string {
  return `@${mentionDisplayName(user)}`;
}

export function matchMentionUsers(users: MentionUser[], query: string) {
  const q = query.trim().toLowerCase();
  if (!q) return users;
  return users.filter((u) => {
    const full = mentionDisplayName(u).toLowerCase();
    const emailLocal = u.email?.split('@')[0].toLowerCase() ?? '';
    return (
      full.includes(q)
      || u.first_name.toLowerCase().includes(q)
      || u.last_name.toLowerCase().includes(q)
      || emailLocal.includes(q)
    );
  });
}

export function extractMentionedUserIds(content: string, users: MentionUser[]): number[] {
  const ids = new Set<number>();
  for (const user of users) {
    if (content.includes(mentionToken(user))) {
      ids.add(user.id);
    }
  }
  return [...ids];
}

const MENTION_SPLIT = /(@[\w]+(?:\s[\w]+)?)/g;

export function splitCommentMentions(content: string): string[] {
  return content.split(MENTION_SPLIT).filter((part) => part.length > 0);
}

export function isMentionPart(part: string): boolean {
  return part.startsWith('@') && part.length > 1;
}
